### PR TITLE
feat(#8461): add contact summary client side error log

### PR DESCRIPTION
**Pull request by jonathanbataire**
# Description

medic/cht-core#8461

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* __CHT_CORE_COMPOSE_URL__
* __COUCH_SINGLE_COMPOSE_URL__
* __COUCH_CLUSTER_COMPOSE_URL__

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by garethbowen** [comment]
@jonathanbataire Is this ready for review?

**Comment by jonathanbataire** [comment]
> @jonathanbataire Is this ready for review?

@garethbowen this is ready for review

**Comment by garethbowen** [comment]
@jkuester As you reviewed the work this builds on, would you mind reviewing please?

The builds that this is waiting on have been renamed so it's not gone green. @jonathanbataire can either rebase, or I can use my admin perms to bypass the check.

File: webapp/src/ts/services/contact-summary.service.ts
Comment (jkuester) [inline webapp/src/ts/services/contact-summary.service.ts:110]: I am hesitant to change this here because it feels like proper behavior for the service to throw an error here. Also, this service is used by more than just the contacts component (e.g. by the [Forms service](https://github.com/medic/cht-core/blob/ee6b24ae719740c8a768250ef078ccce772d10cf/webapp/src/ts/services/form.service.ts#L142)).  

The cleanest alternative I have found is to update the `contacts.effect.ts` and add this catch block to the `loadContactSummary` function (when calling `this.contactSummaryService.get`):

```js
     .catch(err => {
       this.contactsActions.setContactsLoadingSummary(false);
       this.contactsActions.updateSelectedContactSummary(err);
       throw err;
     });
```

Of course, to surface useful exception information, we do not want to be burying the original exception here in the `_get` call.  I think we should just re-throw the original error instead of throwing this `Configuration Error`...

```suggestion
      throw error;
```

File: webapp/src/ts/modules/contacts/contacts-content.component.ts
Comment (jkuester) [inline webapp/src/ts/modules/contacts/contacts-content.component.ts:183]: Based on my comment in the `contact-summary.service`, I think we could update this check to just be:

```suggestion
        if (summary instanceof Error){
          this.summaryErrorStack = summary.stack;
```

@latin-panda I would appreciate your wisdom on this situation. Maybe there is a better way to do this that I have missed.  Basically, we are subscribed to the store here to get the contact summary. Now we want to react (by setting the `summaryErrorStack`) if there is an error getting the contact summary.  The problem is that I cannot seem to figure out a clean way to do any actual error handling here. AFAICT, this code will only be triggered when calling `contactsActions.updateSelectedContactSummary`. Currently, I have suggested just passing the error object instead of the contact summary, but it feels like there has to be a better way....

File: webapp/tests/karma/ts/modules/contacts/contacts-content.component.spec.ts
Comment (jkuester) [inline webapp/tests/karma/ts/modules/contacts/contacts-content.component.spec.ts:1]: Can you include a test here that verifies `component.errorStack` is populated as expected when an Error is returned for the contact summary?

File: tests/e2e/default/contacts/contact-details.wdio-spec.js
Comment (jkuester) [inline tests/e2e/default/contacts/contact-details.wdio-spec.js:71]: Do we need this outer `after` function if we are already resetting the settings in the inner after for `Contact summary error`?

**Review by jkuester — CHANGES_REQUESTED** [review]
Most excellent!  I have a small concern about the error flow which I have suggested a fix for (and mentioned Jenni to see if she has any better ideas...).  Other than that, this is pretty much ready to go!

**Comment by jkuester** [comment]
@tatilepizs this is in good shape.  The e2e test is already done!  Can you have a look and let us know if you have any suggestions/concerns?

**Review by tatilepizs — APPROVED** [review]
It looks good to me! 
Thank you @jonathanbataire for including the e2e test ✨

**Comment by garethbowen** [comment]
@jkuester It looks like you're happy with this change but GH still thinks you've requested changes. If you're happy, can you please hit the big merge button, and close the issue?

**Comment by jkuester** [comment]
@garethbowen I am good with the overall behavior of the code, but I am still a bit concerned about unintended consequences coming from the [change to the error flow](https://github.com/medic/cht-core/pull/8454#discussion_r1330613438).  Was hoping to hear from @latin-panda (or another Angular expert) [if there was a better way](https://github.com/medic/cht-core/pull/8454#discussion_r1330618777) to do error propagation when subscribed to a `store`....

**Comment by garethbowen** [comment]
@latin-panda Josh hoped you could have a look here and see if this is the right way to solve this issue so I've tagged you for review - hope that's ok?

**Comment by latin-panda** [comment]
Apologies, this skipped my radar - distracted with the Nairobi trip last year. I'm going to take a look and comment here.

File: webapp/src/ts/modules/contacts/contacts-content.component.ts
Comment (latin-panda) [reply webapp/src/ts/modules/contacts/contacts-content.component.ts:183]: > The problem is that I cannot seem to figure out a clean way to do any actual error handling here. AFAICT, this code will only be triggered when calling contactsActions.updateSelectedContactSummary. Currently, I have suggested just passing the error object instead of the contact summary, but it feels like there has to be a better way.

@jkuester @jonathanbataire Yeah, it's kind of weird persisting errors in the Store. The error handling is cumbersome because we are [using effects](https://github.com/medic/cht-core/blob/3c3accebdfba8db7ad2c5261eeb75da79e7ad56c/webapp/src/ts/effects/contacts.effects.ts#L72) to load the contact information which is a multi-step process. 
IMO, it's better to do the contact load process in the `contact.service.ts`, and let the component (contact-content.controler.ts) receive and control the result; if it's successful, the component persists the contact info in the Store. If it fails, then display the error. However, this is a bigger refactor and perhaps out of scope for this feature. 
For the time being, adding a catch [here](https://github.com/medic/cht-core/blob/3c3accebdfba8db7ad2c5261eeb75da79e7ad56c/webapp/src/ts/effects/contacts.effects.ts#L163), as store using `this.contactsActions.updateSelectedContactSummary(summary)` seems enough as temporary solution. 
```
return this.contactSummaryService
      .get(selected.doc, selected.reports, selected.lineage, selected.targetDoc)
      .catch(error => {
        return { error };
      })
      .then(summary => {
        return this
          .verifySelectedContactNotChanged(contactId)
          .then(() => {
            this.contactsActions.setContactsLoadingSummary(false);
            return this.contactsActions.updateSelectedContactSummary(summary);
          });
      });
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Please find my feedback below

**Comment by jonathanbataire** [comment]
thanks @latin-panda following up on this

File: tests/e2e/default/contacts/contact-details.wdio-spec.js
Comment (jonathanbataire) [reply tests/e2e/default/contacts/contact-details.wdio-spec.js:71]: yes otherwise it throws an error `A previous test did not call revertSettings` when running error test.

File: webapp/src/ts/effects/contacts.effects.ts
Comment (jonathanbataire) [inline webapp/src/ts/effects/contacts.effects.ts:96]: @latin-panda catching the error inside `loadContactSummary` meant that this catch block here wouldnt execute
so to kinda harmonize everything i decided to updatedSummary in this catch block. please let me know what you think about this. thanks

Comment (latin-panda) [reply webapp/src/ts/effects/contacts.effects.ts:96]: Thanks for the update!
This function has many processes going on, and putting it here will retrieve all the errors of any of those processes, which is also not in the scope of this ticket. 

You can try this: 
- add a catch block in the loadContactSummary, 
- inside that block call updateSelectedContactSummary
- then throw the error again

Like this, it will keep the usage of `updateSelectedContactSummary` inside the same function; that's good because it keeps it contained.  And by throwing the error, it will execute the catch block in this `selectContact` effect. 

```javascript

private loadContactSummary(contactId, trackName) {
    const trackPerformance = this.performanceService.track();
    const selected = this.selectedContact;
    return this.contactSummaryService
      .get(selected.doc, selected.reports, selected.lineage, selected.targetDoc)
      .then(summary => {
        return this
          .verifySelectedContactNotChanged(contactId)
          .then(() => {
            this.contactsActions.setContactsLoadingSummary(false);
            return this.contactsActions.updateSelectedContactSummary(summary);
          });
      })
      .catch(error => {
         this.contactsActions.updateSelectedContactSummary({ errorStack: error.stack });
         throw error;
      })
      .finally(() => {
        trackPerformance?.stop({ name: [ ...trackName, 'load_contact_summary' ].join(':') });
      });
  }
```

**Review by latin-panda — CHANGES_REQUESTED** [review]
Thanks for the update! I've added a suggestion inline

Tatiana is unavailable to review this. @lorerod do you have availability to have a quick look at the e2e?

File: webapp/src/ts/effects/contacts.effects.ts
Comment (jonathanbataire) [reply webapp/src/ts/effects/contacts.effects.ts:96]: :) awesome

**Review by latin-panda — APPROVED** [review]
The revision looks good to me.
Just waiting for @jkuester and @lorerod feedback before we can close this work.
Thanks Jonathan!

**Comment by latin-panda** [comment]
I've updated the PR title to match the format of `feat(#8461): ......`

**Review by jkuester — APPROVED** [review]
Lets go! :rocket:

**Review by lorerod — CHANGES_REQUESTED** [review]
Thank you, @jonathanbataire, for including the unit tests and end-to-end tests. I have a minor suggestion.

File: tests/page-objects/default/contacts/contacts.wdio.page.js
Comment (latin-panda) [reply tests/page-objects/default/contacts/contacts.wdio.page.js:382]: Hi @jonathanbataire, Have you had the chance to look at @lorerod feedback?

Comment (jonathanbataire) [reply tests/page-objects/default/contacts/contacts.wdio.page.js:382]: sorry totally missed this 
@lorerod the selectors for the errorlog are different in both. thats the only diff
`const errorLogSelector = () => $('.item-content error-log');` in contacts  vs `const errorLog = () => $('.page error-log');` in analytics
im pretty divided on how to go about this. isnt the difference significant enough to separate the 2? idk
lmk what you think.

Comment (lorerod) [reply tests/page-objects/default/contacts/contacts.wdio.page.js:382]: Hi @jonathanbataire @latin-panda, what do you think about moving the method to the `common.wdio.page.js` and making it receive the `errorLog` selector as a parameter?

**Review by lorerod — APPROVED** [review]
Hi @jonathanbataire @latin-panda, You can go ahead and merge this PR. My suggestion is just to improve the code, not to block you. Sorry. Thanks.

File: tests/page-objects/default/contacts/contacts.wdio.page.js
Comment (tatilepizs) [reply tests/page-objects/default/contacts/contacts.wdio.page.js:382]: Hi all,
I don't know if the error text is the same for contacts, analytics, and tasks, but if it is, we can use a selector by text, and it will be the same method for all three places.

@jonathanbataire look [this example](https://github.com/medic/cht-core/blob/master/tests/page-objects/upgrade/upgrade.wdio.page.js#L4-L6) of the upgrade page object file, in in which we use the selector by text 🙂 

Let me know if I can help..

Comment (jonathanbataire) [reply tests/page-objects/default/contacts/contacts.wdio.page.js]: Hello @tatilepizs @lorerod this has been updated getErrorLog moved to `commompage`

Comment (tatilepizs) [reply tests/page-objects/default/contacts/contacts.wdio.page.js]: That is great, thank you @jonathanbataire for making the change and updating all the instances ⭐ this really help us to keep our code clean and organized 👍

**Review by tatilepizs — APPROVED** [review]
Looks great, thank you for the improvements to the testing code ⭐

**Comment by latin-panda** [comment]
@jonathanbataire, the CI errors are weird, can you try downloading the latest from `master` branch? and maybe they get fixed

**Comment by latin-panda** [comment]
@garethbowen have you seen this error before? 
```
 Error: the string "Error response from daemon: manifest for medicmobile/cht-haproxy:4.10.0-8039-contactsummary-log.9772146005 not found: manifest unknown: manifest unknown\n" was thrown, throw an Error :)
```
The PR changes don't seem related to this, and the branch has been updated with the latest from `master`.

**Comment by latin-panda** [comment]
@garethbowen I see this `feat(#9227): add validation function for luhn algorithm` from ChinHairSaintClair has the same problem, so maybe it's because it's coming from a contributor

**Comment by garethbowen** [comment]
@latin-panda I think you're right. Please raise an issue to fix the k3d tests for external contributors.

For now, those builds aren't required so we can ignore them. There are two other problems - one test (probably flaky) and SonarCloud isn't responding.

**Comment by latin-panda** [comment]
I don't see any option to restart the sonar from my user ([sonar link](https://sonarcloud.io/summary/new_code?id=medic_cht-core&pullRequest=8454))

Screenshot 2024-07-03 at 12 57 28 PM

@jonathanbataire, you can try pushing a commit, like a space or something small to see if Sonar picks it up

**Comment by garethbowen** [comment]
@latin-panda Let's get this merged. I know the k3d builds are failing because of a known issue. Do you know what the issue is with sonar? I wonder if it's also not working for external contributors. Can you investigate and let me know?

If everything's fine then I can merge it bypassing branch protections - just let me know!

**Comment by latin-panda** [comment]
@garethbowen It seems like my user in Sonar UI is limited ([sonar link](https://sonarcloud.io/summary/new_code?id=medic_cht-core&pullRequest=8454)). I don't see any action I can click to help passing this PR. 
It's stuck in an old commit and not receiving the new ones. 

Another contributor seems to have the problem `add validation function for luhn algorithm` I cannot even find the PR [in sonar](https://sonarcloud.io/project/pull_requests_list?id=medic_cht-core). Perhaps something to do in project config of Sonar for Anonymous users.

**Comment by garethbowen** [comment]
@latin-panda Yes I think that's right. Can you run sonar locally to see if this PR passes?

**Comment by jkuester** [comment]
:thinking: I don't think we are having Sonar problems for all contributor PRs, since I think it was working okay for the Draw Widget PR.  I do notice that the last time Sonar analyzed this PR was on the last merge-commit from `master`. I wonder if something has gotten funky with how Sonar is seeing the commit graph here.  My suggestion would be to try and update the branch with a fresh merge-commit from `master` and see if Sonar picks things back up from there...

**Comment by jkuester** [comment]
Hmm, rats. That does not seem to have helped.  :sweat: I also do not see any  way to manually trigger the check from the SonarCloud side of things.

One complicating factor is that it looks like this PR was opened before we actually [enabled Sonar on this repo](https://github.com/medic/cht-core/commit/fe3f64d2729b00024a3775ed0cb3fd091e8ccffb).  That may be one reason why this is going so weird....  Either way, @garethbowen, it should be safe to just override the check and merge this PR.  I created a [new branch/PR](https://github.com/medic/cht-core/pull/9251) with the same HEAD commit and it passed the Sonar checks fine. So, no concern here.

**Comment by garethbowen** [comment]
@jkuester Thanks for confirming.

@jonathanbataire Thanks for your patience! This is now merged.

**Pull request merged**
